javalib/src/main/scala/java/lang/Float.scala

```diff
@@ -293,11 +293,75 @@ object Float {
       zUp
   }
 
-  @inline private def multiplyBy10Pow(v: BigInteger, e: Int): BigInteger =
-    v.multiply(BigInteger.TEN.pow(e))
+  /** An implementation of big integer arithmetics that we need in the above method. */
+  private sealed abstract class BigIntImpl {
+    type Repr
+
+    def fromString(str: String): Repr
+
+    /** Creates a big integer from a `Long` that needs at most 53 bits (unsigned). */
+    def fromUnsignedLong53(x: scala.Long): Repr
+
+    def multiplyBy2Pow(v: Repr, e: Int): Repr
+    def multiplyBy10Pow(v: Repr, e: Int): Repr
+
+    def compare(x: Repr, y: Repr): Int
+  }
+
+  private object BigIntImpl {
+    /** Get the best available implementation of big integers for the given platform.
+     *
+     *  If JS bigint's are supported, use them. Otherwise fall back on
+     *  `java.math.BigInteger`.
+     *
+     *  We need a `linkTimeIf` here because the JS bigint implementation uses
+     *  the `**` operator, which does not link when `esVersion < ESVersion.ES2016`.
+     */
+    @inline def getImpl(): BigIntImpl = {
+      import scala.scalajs.LinkingInfo._
+      linkTimeIf[BigIntImpl](esVersion >= ESVersion.ES2020) {
+        JSBigInt
+      } {
+        JBigInteger
+      }
+    }
+
+    private object JSBigInt extends BigIntImpl {
+      type Repr = js.BigInt
 
-  @inline private def multiplyBy2Pow(v: BigInteger, e: Int): BigInteger =
-    v.shiftLeft(e)
+      @inline def fromString(str: String): Repr = js.BigInt(str)
+
+      /* The 53-bit restriction allows us to take twice 31 bits, which avoids
+       * the issue of considering the bit 31 as a sign bit in the rhs of the |.
+       * It also guarantees we can safely build a `Double` out of it.
+       */
+      @inline def fromUnsignedLong53(x: scala.Long): Repr =
+        js.BigInt((x >>> 31).toInt.toDouble * 2147483648.0 + (x.toInt & 0x7fffffff).toDouble)
```

Comment:
I couldn’t see the difference from `js.BigInt(x.toDouble)`. Sorry if I misunderstood anything! 
Did you mean something like this?

```suggestion
        js.BigInt((x >>> 31).toInt.toDouble) * js.BigInt(2147483648L) + js.BigInt((x.toInt & 0x7fffffff).toDouble)
```

Reply:
You're not missing anything. It is equivalent to `js.BigInt(x.toDouble)`. In JS `x.toDouble` needs to handle the negative cases, though, which is a bit more involved, and which I wanted to avoid. But that was probably premature optimization. I switched to `x.toDouble`.